File: cpp/stack&queue/Queue.cpp
#include <fstream>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

class Node {
public:
  Node *next;
  std::string value;

  Node() : next(nullptr) {}
};

class Queue {
public:
  Node *front;
  Node *rear;

  Queue() : front(nullptr), rear(nullptr) {}

  bool isEmpty() const { return front == nullptr; }

  void enqueue(const std::string &value) {
    Node *node = new Node();
    node->value = value;
    node->next = nullptr;
    if (isEmpty()) {
      front = rear = node;
    } else {
      rear->next = node;
      rear = node;
    }
  }

  std::string dequeue() {
    if (isEmpty()) {
      std::cout << "Index out of range" << std::endl;
      return "";
    }
    Node *node = front;
    front = front->next;
    if (front == nullptr) {
      rear = nullptr;
    }
    std::string value = node->value;
    delete node;
    return value;
  }

  void swap(int index1, int index2) {
    if (index1 == index2) {
      return;
    }
    if (index1 > index2) {
      std::swap(index1, index2);
    }
    Node *node1 = front;
    Node *node2 = front;
    for (int i = 0; i < index1 && node1; ++i) {
      node1 = node1->next;
    }
    for (int i = 0; i < index2 && node2; ++i) {
      node2 = node2->next;
    }
    if (node1 && node2) {
      std::swap(node1->value, node2->value);
    }
  }

  std::string get(int index) {
    Node *node = front;
    for (int i = 0; i < index && node; ++i) {
      node = node->next;
    }
    if (node) {
      return node->value;
    } else {
      std::cout << "Index out of range" << std::endl;
      return "";
    }
  }
};

Queue csv_to_queue(const std::string &filename) {
  Queue queue;
  std::ifstream csvfile(filename);
  std::string line;
  std::vector<std::string> lines;

  while (std::getline(csvfile, line)) {
    queue.enqueue(line);
  }

  return queue;
}

void queue_to_csv(const Queue &queue, const std::string &filename) {
  std::ofstream csvfile(filename);
  Node *current = queue.front;
  while (current) {
    csvfile << current->value << std::endl;
    current = current->next;
  }
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    std::cout << "Usage: " << argv[0] << " <operation> <args...>" << std::endl;
    return 1;
  }

  Queue queue = csv_to_queue("List.csv");

  std::string operation = argv[1];

  if (operation == "swap" && argc == 4) {
    queue.swap(std::stoi(argv[2]), std::stoi(argv[3]));
  } else if (operation == "enqueue" && argc == 3) {
    queue.enqueue(argv[2]);
  } else if (operation == "dequeue") {
    queue.dequeue();
  } else if (operation == "get" && argc == 3) {
    std::cout << queue.get(std::stoi(argv[2])) << std::endl;
  } else {
    std::cout << "Invalid operation" << std::endl;
  }

  queue_to_csv(queue, "List.csv");

  return 0;
}

File: python/stack&queue/Queue.py
from __future__ import annotations
from typing import Any
import csv
import argparse


class Node:
    value: Any
    next: Node | None

    def __init__(self) -> None:
        self.value: Any = None
        self.next: Node | None = None


class Queue:
    def __init__(self) -> None:
        self.front: Node | None = None
        self.rear: Node | None = None

    def __iter__(self) -> Any:
        node = self.front
        while node:
            yield node.value
            node = node.next

    def __str__(self) -> str:
        return str(list(self))

    def isEmpty(self) -> bool:
        return self.front is None

    def enqueue(self, value: Any) -> None:
        node = Node()
        node.value = value
        if self.rear is None:
            self.front = self.rear = node
        else:
            self.rear.next = node
            self.rear = node

    def dequeue(self) -> Any:
        if self.isEmpty():
            print("Index out of range")
            return
        node = self.front
        if node:
            self.front = node.next
        if self.front is None:
            self.rear = None
        if node:
            return node.value

    def swap(self, index1, index2):
        if index1 == index2:
            return
        if index1 > index2:
            index1, index2 = index2, index1
        node1 = self.front
        node2 = self.front
        for _ in range(index1):
            if node1:
                node1 = node1.next
        for _ in range(index2):
            if node2:
                node2 = node2.next
        if node1 and node2:
            node1.value, node2.value = node2.value, node1.value

    def get(self, index):
        node = self.front
        for _ in range(index):
            if node:
                node = node.next
        if node:
            return node.value
        else:
            print("Index out of range")


def csv_to_queue(filename: str) -> Queue:
    queue = Queue()
    with open(filename, newline="") as csvfile:
        reader = csv.reader(csvfile)
        for _ in list(*reader):
            queue.enqueue(_)
    return queue


def queue_to_csv(queue: Queue, filename: str) -> None:
    with open(filename, "w", newline="") as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(iter(queue))


def main():
    parser = argparse.ArgumentParser(
        description="Perform operations on a Queue Data Structure from a CSV file."
    )
    parser.add_argument(
        "operation",
        choices=["swap", "enqueue", "dequeue", "get"],
        help="Operation to perform on the Queue data structure",
    )
    parser.add_argument(
        "args", nargs="*", help="Additional arguments for the operation"
    )

    args = parser.parse_args()
    stack = csv_to_queue("List.csv")

    match args.operation:
        case "swap":
            stack.swap(int(args.args[0]), int(args.args[1]))
        case "enqueue":
            stack.enqueue(args.args[0])
        case "dequeue":
            stack.dequeue()
        case "get":
            print(stack.get(int(args.args[0])))
        case _:
            print("Invalid operation")

    queue_to_csv(stack, "List.csv")


if __name__ == "__main__":
    main()
